File: Napakalaki Java/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;
import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author Antonio
 */
public class Napakalaki {
    
    //////////////////////////////////////////////////
    /////////////////////////////////////////////
    private static Napakalaki game = null;
    private Monster currentMonster = null;
    private CardDealer dealer = null;
    private ArrayList<Player> players = null;
    private Player currentPlayer = null;
            
    
    private Napakalaki(){
        
        currentMonster = null;
        currentPlayer = null;
        dealer =  CardDealer.getInstance();
        players = new ArrayList();
       
    
    }
    
    private void initPlayers(ArrayList<String> names){
        for (String name : names) {
            players.add(new Player(name));
        }      
    }
    
    private Player nextPlayer(){
        Player next = null;
        Random rand = new Random();
        if(currentPlayer == null) {
            next = players.get(rand.nextInt(players.size()));
        } else {
            for (int i=0; i<players.size();i++)
                if (players.get(i)==currentPlayer)
                    next = players.get((i+1)%players.size());
        }
        currentPlayer = next;
        return next;
    }
    
    private boolean nextTurnAllowed(){
        boolean allowed;
        if(currentPlayer != null)
            allowed = currentPlayer.validState();
        else 
            allowed = true;
        
        return allowed;
    }
    
    private void setEnemies(){
        boolean asignado = false;
        int aux;
        int pos;
        if(players.size() == 1){
            Player p = players.get(0);
            p.setEnemy(p);
        }else{
            for (Player  player : players) {
                asignado = false;
                while (!asignado) {
                    aux = (int) (Math.random()*(players.size()));
                    pos = players.indexOf(player);        
                    if (pos != aux) {
                        player.setEnemy(players.get(aux));
                        asignado = true;
                    }
                }
            }
        }
    }
    
    ///////////////////////////////////////
    ///////////////////////////////////////
    public static Napakalaki getInstance(){
        if(game == null){
            game = new Napakalaki();
        }
        return game;
    }
    
    public CombatResult developCombat(){
        CombatResult resultado;
        
        resultado = currentPlayer.combat(currentMonster);
        dealer.giveMonsterBack(currentMonster);
        
        if(resultado == CombatResult.LOOSEANDCONVERT) {
            Cultist sectario = dealer.nextCultist();
            CultistPlayer cultistP = new CultistPlayer(currentPlayer, sectario);
            int indice = players.indexOf(currentPlayer);
            players.set(indice, cultistP); 
            for(Player p:players) {
                if(p.getEnemy() == currentPlayer)
                    p.setEnemy(cultistP);
            }
            
            currentPlayer = cultistP;

        }
      
        
        return resultado;
        
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        
        ArrayList<Treasure> aux = currentPlayer.getVisibleTreasures();
       
        for (Treasure  tesoro : treasures){
            
                currentPlayer.discardVisibleTreasure(tesoro);
                dealer.giveTreasureBack(tesoro);
        }
        
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        
        ArrayList<Treasure> aux = currentPlayer.getHiddenTreasures();
     
        for (Treasure  tesoro : treasures){
       
                currentPlayer.discardHiddenTreasure(tesoro);
                dealer.giveTreasureBack(tesoro);
            
        }
        
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        
        for (Treasure  tesoro : treasures){
            currentPlayer.makeTreasureVisible(tesoro);
        }
        
    }
    
    public void initGame(ArrayList<String> nombres){
        initPlayers(nombres);      
        setEnemies();
        dealer.initCards();
        nextTurn();              
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
        
    }
    
    public boolean nextTurn(){
        boolean continuar = nextTurnAllowed();
        
        if(continuar){
            currentMonster = dealer.nextMonster();  
            currentPlayer = nextPlayer();
            if(currentPlayer.isDead()){
                currentPlayer.initTreasures();
            }    
        }
        return continuar;
    }
    
    public boolean endOfGame(CombatResult result){
        return result == CombatResult.WINGAME;
    }
       
}

File: Napakalaki Ruby/Napakalaki.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


require_relative 'Player.rb'
require_relative 'CultistPlayer.rb'
require_relative 'Cultist.rb'
require_relative 'CardDealer.rb'
require_relative 'NumericBadConsequence.rb'
require_relative 'DeathBadConsequence.rb'
require_relative 'SpecificBadConsequence.rb'

module NapakalakiGame

class Napakalaki
  
  attr_reader :current_player, :players, :current_monster
  
  @@instance = nil
  private_class_method :new
  
private

  def initialize ()
    @players = Array.new
    #@players << Player.new("jugador0")
    @current_player = nil
    @dealer = CardDealer.instance()
    @current_monster = nil
  end
  
  def initPlayers(nombres)
    i=0;
    
    while(i<nombres.length)
      aux = Player.new(nombres.at(i))
      i+=1
      @players << aux
    end
  end
  
  
  def nextPlayer()
    
    if(@current_player == nil)
      indice = rand(@players.length)
      jugador_aux = @players.at(indice) 
      @current_player = jugador_aux
    else
      indice = (@players.index(@current_player) +1)%@players.length
      jugador_aux = @players.at(indice)
      @current_player = jugador_aux  
    end
    
    return @current_player
  end
  
  def nextTurnAllowed()
    permitido = true
    if(@current_player != nil)
    permitido = @current_player.validState
    end
    return permitido
  end
  
  def setEnemies()
    i=0
    if(@players.length == 1)
      p = @player.get(0)
      p.setEnemy(p)
    else
      while(i<@players.length)
        indice = i 
        while(indice == i)
          indice = rand(@players.length)
        end
        (@players.at(i)).setEnemy(@players.at(indice))
        i+=1
      end
    
    end
  end
  
public

  def self.instance()
    @@instance = new if (@@instance == nil)
    return @@instance
  end
  
  def developCombat()
    combat_result = @current_player.combat(@current_monster)
    @dealer.giveMonsterBack(@current_monster)
    if(combat_result == CombatResult::LOSEANDCONVERT)
      carta_aux = @dealer.nextCultist
      jugador_aux = CultistPlayer.new(@current_player,carta_aux)
      i = @players.find_index(@current_player)
      @players[i] = jugador_aux
      
      @players.each do |x|
        if(x.getEnemy == @current_player)
          x.setEnemy(jugador_aux)
        end
      end
      @current_player = jugador_aux
      
    end
    return combat_result
  end
  
  def discardVisibleTreasures (treasures)
    for i in treasures 
      @current_player.discardVisibleTreasure(i)
      @dealer.giveTreasureBack(i)
    end
  end
  
  def discardHiddenTreasures (treasures)
    for i in treasures 
      @current_player.discardHiddenTreasure(i)
      @dealer.giveTreasureBack(i)
    end
  end
  
  def makeTreasuresVisible(tesoros)
    for i in tesoros
      @current_player.makeTreasureVisible(i)
    end
  end
  
  def initGame (names)
    initPlayers(names)
    setEnemies
    @dealer.initCards
    nextTurn
    
  end
  
  def nextTurn ()
    
    state_ok = nextTurnAllowed
    if (state_ok)
      @current_monster = @dealer.nextMonster
      @current_player = nextPlayer
      dead = @current_player.isDead
      @current_player.initTreasures if(dead)
    end
    return state_ok
  end
  
  def endOfGame (resultado)
    fin = false
    
    if(resultado == CombatResult::WINGAME)
      fin = true
    end  
     
    return fin
  end
  
  def getCurrentPlayer()
    
    return @current_player
  end
  
  def getCurrentMonster()
    
    return @current_monster
  end
end


end


